File: benchmarks/bench.py
# KamiPython benchmark suite — the same file runs under CPython and kamipy.
# Prints "name,seconds" lines so run_benchmark.py can build a comparison table.
import time


def integer_loop(n):
    total = 0
    for i in range(n):
        total = (total + i * 3) % 1000000007
    return total


def math_test(n):
    acc = 0.0
    for i in range(n):
        acc = acc + i * 0.5 - acc / 3.0
    return acc


def fib(n):
    if n < 2:
        return n
    return fib(n - 1) + fib(n - 2)


def string_test(n):
    s = ""
    for i in range(n):
        s += "x"
    return len(s)


def bench(name, fn, arg):
    t0 = time.perf_counter()
    r = fn(arg)
    t1 = time.perf_counter()
    print(name + "," + str(t1 - t0) + "," + str(r))


bench("integer_loop", integer_loop, 100000000)
bench("math", math_test, 100000000)
bench("recursion", fib, 32)
bench("string_test", string_test, 1000000)

File: benchmarks/bench.go
// Go reference implementation of benchmarks/bench.py (same algorithms).
package main

import (
	"fmt"
	"strings"
	"time"
)

func integerLoop(n int64) int64 {
	var total int64 = 0
	for i := int64(0); i < n; i++ {
		total = (total + i*3) % 1000000007
	}
	return total
}

func mathTest(n int64) float64 {
	acc := 0.0
	for i := int64(0); i < n; i++ {
		acc = acc + float64(i)*0.5 - acc/3.0
	}
	return acc
}

func fib(n int64) int64 {
	if n < 2 {
		return n
	}
	return fib(n-1) + fib(n-2)
}

// Idiomatic Go: += on strings is O(n^2), so Go programs use strings.Builder.
// (KamiPython's escape analysis gives plain `s += "x"` the same behavior.)
func stringTest(n int64) int {
	var b strings.Builder
	for i := int64(0); i < n; i++ {
		b.WriteString("x")
	}
	return len(b.String())
}

func bench(name string, f func() interface{}) {
	t0 := time.Now()
	r := f()
	dt := time.Since(t0).Seconds()
	fmt.Printf("%s,%v,%v\n", name, dt, r)
}

func main() {
	bench("integer_loop", func() interface{} { return integerLoop(100000000) })
	bench("math", func() interface{} { return mathTest(100000000) })
	bench("recursion", func() interface{} { return fib(32) })
	bench("string_test", func() interface{} { return stringTest(1000000) })
}
